Updates to python style guide

lainets opened the pull request:
This PR exists to update and discuss the python style guide as some of the rules are ambiguous or out dated.

The first thing to update that was already agreed upon in a separate meeting is to allow interchangeable usage of single and double quotes.

lainets committed:
Update python quote style to whatever is natural

Double or single quotes may be used interchangeably but there are
some restrictions in cases where strings themselves contain quotes
to make them readable.

lainets commented:
> Try to limit lines to **79** characters, and don't exceed **99** characters.
> ...
> In addition, use a longer line (not exceeding 99 characters), if the code is more readable that way!

This is very short. I know this comes from PEP-8 but I do not think this short lines are necessary with modern machines. This rule is broken a lot in the code base anyway. Something like 119 characters would work much better in my opinion. Based on the upvotes at https://stackoverflow.com/questions/88942/why-does-pep-8-specify-a-maximum-line-length-of-79-characters, most people agree (though the upvotes might be very biased). Though, I'm not that strongly opinionated about this.

> Typically, the code should prefer format method. Remember, when using lazily evaluated strings, then the formatting should be lazy too. For example, django gettext returns a lazy object, hence format_lazy should be used with it.

I would argue f-string should be preferred. The whole reasoning for them was a concise and readable way to format strings [PEP 498](https://www.python.org/dev/peps/pep-0498/). Of course, format and format_lazy should be used when f-strings aren't viable.

>  5. public helpers
> 6. public classes
> 7. private helpers
> 8. private classes
 
 Having the private helpers and classes after public ones will cause issues with type hints pre 3.10. This _can_ be solved by adding `from __future__ import annotations` but that might break some libraries (though none of them are probably used by A+ at the moment). I think it is also more common to have private members before public ones in general. E.g. the standard library seems to do it that way: [example file](https://github.com/python/cpython/blob/main/Lib/http/cookiejar.py). I would also argue that if a private class/function is only used by a single class, it should be right above that class to group them together.

lainets committed:
surround (with blank lines) -> separate

Separate seems like the right word to use here

markkuriekkinen commented:
Line length: I can also support 119-character long code lines.

String format method: we couldn't support f-strings in Python 3.5, but A+ has already moved to newer Python, thus f-strings are now possible (and more performant). However, Django's lazy (translated) strings only support the Django function `format_lazy`. There is no other way to format them correctly. 

Private members before public API functions: fine by me.

markkuriekkinen reviewed (approved):
This is good. Do you want to add more changes or shall we merge the PR?

skulonen commented:
Interestingly, [this page](https://apluslms.github.io/contribute/styleguides/python/) says 99 is the absolute maximum line length, but [this page](https://apluslms.github.io/contribute/styleguides/) says 119.

lainets commented:
I'll add some changes based on my previous comment.

lainets committed:
change python hard line limit to 119

See PR #9
TL;DR modern machines support this well and this is also used on the
style guide index page

lainets committed:
prefer f-string over format

skulonen commented:
According to the style guide, this is how multiline function declarations should be indented:
```python
def foobar(
        argument1,
        argument2,
        keyword1=None,
        keyword2=None, # remember the comma
        ): # here this seems to be part of the arguments, which is good
    if keyword1 is None:
```
But what about classes and long inheritance lists? The style guide says nothing about them. It seems that the codebase currently uses this style:
```python
class SomeClass(BaseClass1,
                BaseClass2,
                BaseClass3):
```

lainets committed:
private members before public ones in python

lainets commented:
I would prefer to use the same style as for functions because the syntax is essentially the same but the negatives of changing the style would probably out weight the positives of using the same style for both classes and functions.

markkuriekkinen commented on contribute/styleguides/python/index.md:
```diff
@@ -283,13 +283,19 @@ total_cost = (
 10. the `if __name__ == '__main__'`
 
+Private members should be before public ones because it is a common style, and
+it is required for type hints to work without future import pre 3.10.
```
Does "3.10" mean Python 3.10?

lainets replied:
Yes. Should I specify python?

lainets committed:
specify 3.10 refers to python

lainets commented:
This ready for merging from my side.
